src/bin/common/solving_loop.ml

```diff
@@ -356,6 +409,53 @@ let main () =
     State.create_key ~pipe:"" "is_incremental"
   in
 
+  (* Set to true when a query is performed, states for the following SMT
+     instructions that they are working on an environment in which some formulae
+     have been decided and must be reverted back to the one before the check-sat
+     before starting to assert again. *)
+  let is_decision_env : bool State.key =
+    State.create_key ~pipe:"" "is_decision_env"
+  in
+
+  (* Key used only in handle_stmt_pop_reinit, registers the commands executed
+     to save them them if a push command is executed. *)
+  let current_path_key : stmt list State.key =
+    State.create_key ~pipe:"" "current_path"
+  in
+
+  (* Key used only in handle_stmt_pop_reinit, saves the commands executed
+     when a push is executed and replays them when a pop is met. *)
+  let pushed_paths_key : stmt list Vec.t State.key =
+    State.create_key ~pipe:"" "pushed_paths"
+  in
+
+  (* Before each query, we push the current environment. This allows to keep a
+     fresh one for the next assertions. *)
+  let internal_push st =
```

Comment:
I think this could use a more explicit name such as `push_before_query` (especially since the `Frontend` module also has an `internal_push` function that behaves very differently).